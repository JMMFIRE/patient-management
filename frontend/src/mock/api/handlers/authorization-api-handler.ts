import {rest} from 'msw'

const testUser = {
    userId: 1,
    email: 'massottoj@envisionpharma.com',
    password: 'test',
    firstName: 'Jacob',
    lastName: 'Massotto',
    role: 'General Doctor'
}
type LoginRequestBody = {
    email: string,
    password: string
}

export const handlers = [
    rest.post<LoginRequestBody>('http://localhost:8080/users/login', (req, res, ctx) => {
        const { email, password } = req.body

        if (email === testUser.email && password === testUser.password) {
            return res(ctx.status(200), ctx.json(testUser))
        }
        else {
            return res(ctx.status(500))
        }
    })
]
