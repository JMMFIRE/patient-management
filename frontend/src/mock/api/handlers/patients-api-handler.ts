import {rest} from 'msw'

let patientList: Patient []

/**
 * Helper method to reset the data assigned to the patientList for testing
 */
export const resetPatientList = () => {
    const initialPatientList: Patient[] = [
        {
            recordNumber: 1,
            firstName: 'Test',
            lastName: 'Patient1',
            dateOfBirth: new Date('2022-05-29T04:00:00.000+00:00'),
            sex: 'male',
            phoneNumber: '+447277713152',
            diagnosis: 'test diagnosis',
            status: 'Awaiting Surgery',
            notes: 'test notes',
            lastAppointment: new Date('2023-01-18T05:00:00.000+00:00'),
            nextAppointment: new Date ('2023-01-20T05:00:00.000+00:00'),
            doctor: {
                userId: 1,
            }
        },
        {
            recordNumber: 2,
            firstName: 'Test',
            lastName: 'Patient2',
            dateOfBirth: new Date('2022-05-29T04:00:00.000+00:00'),
            sex: 'male',
            phoneNumber: '+447277713152',
            diagnosis: 'test diagnosis',
            status: 'Recovered',
            notes: 'test notes',
            lastAppointment: new Date('2023-01-18T05:00:00.000+00:00'),
            nextAppointment: new Date ('2023-01-20T05:00:00.000+00:00'),
            doctor: {
                userId: 1,
            }
        },
        {
            recordNumber: 3,
            firstName: 'Test',
            lastName: 'Patient3',
            dateOfBirth: new Date('2022-05-29T04:00:00.000+00:00'),
            sex: 'male',
            phoneNumber: '+447277713152',
            diagnosis: 'test diagnosis',
            status: 'On Treatment',
            notes: 'test notes',
            lastAppointment: new Date('2023-01-18T05:00:00.000+00:00'),
            nextAppointment: new Date ('2023-01-20T05:00:00.000+00:00'),
            doctor: {
                userId: 1,
            }
        },
    ]

    patientList = initialPatientList
}

export const handlers = [
    rest.get('http://localhost:8080/patients/doctor/:userid', (req, res, ctx) => {
        return res(ctx.status(200), ctx.json((patientList)))
    }),
    rest.post<Patient>('http://localhost:8080/patients/patient', (req, res, ctx) => {
        const newPatient = {...req.body, recordNumber: 11}

        patientList.push(newPatient)
        return res(ctx.status(200))
    }),
    rest.delete('http://localhost:8080/patients/:patientId', (req, res, ctx) => {
        const { patientId } = req.params
        let response

        if (patientId) {
            const patientIndex = patientList.findIndex((patient) => patient.recordNumber?.toString() === patientId)
            if (patientIndex !== -1) {
                patientList.splice(patientIndex, 1)
                response = res(ctx.status(200))
            } else
                response = res(ctx.status(404))
        }
        return response
    })
]
