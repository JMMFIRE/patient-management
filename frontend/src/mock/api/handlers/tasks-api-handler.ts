import {rest} from 'msw'

let taskList: Task[]

type TaskCompletedRequest = {
    completed: boolean
}

/**
 * Helper method to reset the data assigned to the taskList for testing
 */
export const resetTaskList = () => {
    const initialTaskList: Task [] = [
        {
            taskNumber: 6383,
            title: 'Test Task',
            description: '',
            completed: false,
            createdDate: new Date('2023-02-27T05:00:00.000+00:00'),
            dueDate: new Date('2023-02-15T05:00:00.000+00:00'),
            tags: [],
            doctor: {
                userId: 1
            }
        },
        {
            taskNumber: 6384,
            title: 'Test Task 2',
            description: '',
            completed: true,
            createdDate: new Date ('2023-02-27T05:00:00.000+00:00'),
            dueDate: new Date('2023-02-15T05:00:00.000+00:00'),
            tags: ['a', 'test', 'tag'],
            doctor: {
                userId: 1
            }
        }
    ]

    taskList = initialTaskList
}

export const handlers = [
    rest.get('http://localhost:8080/tasks/doctor/:userId', (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(taskList))
    }),
    rest.post<Task>('http://localhost:8080/tasks/task', (req, res, ctx) => {
        const newTask = {...req.body, taskNumber: 11}

        taskList.push(newTask)
        return res(ctx.status(200))
    }),
    rest.delete('http://localhost:8080/tasks/:taskId', (req, res, ctx) => {
        const { taskId } = req.params
        let response

        if (taskId) {
            const taskIndex = taskList.findIndex((task) => task.taskNumber?.toString() === taskId)
            if (taskIndex !== -1) {
                taskList.splice(taskIndex, 1)
                response = res(ctx.status(200))
            } else
                response = res(ctx.status(404))
        }
        return response
    }),
    rest.put<Task>('http://localhost:8080/tasks/:taskId', (req, res, ctx) => {
        const { taskId } = req.params
        const task = req.body
        let response

        if (taskId) {
            const taskIndex = taskList.findIndex((task) => task.taskNumber?.toString() === taskId)
            if (taskIndex !== -1) {
                taskList[taskIndex] = task
                response = res(ctx.status(200))
            } else
                response = res(ctx.status(404))
        }
        return response
    }),
    rest.patch<TaskCompletedRequest>('http://localhost:8080/tasks/:taskId', (req, res, ctx) => {
        const { taskId } = req.params
        const { completed } = req.body
        let response

        if (taskId) {
            const taskIndex = taskList.findIndex((task) => task.taskNumber?.toString() === taskId)
            if (taskIndex !== -1) {
                taskList[taskIndex] = {...taskList[taskIndex], completed}
                response = res(ctx.status(200))
            } else
                response = res(ctx.status(404))
        }
        return response
    })
]
