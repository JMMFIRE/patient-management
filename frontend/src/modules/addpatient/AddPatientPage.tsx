import styles from './AddPatientPage.module.scss'
import 'react-datepicker/dist/react-datepicker.css'
import {useState} from 'react'
import {useNavigate} from 'react-router'
import AddPatientForm from './AddPatientForm/AddPatientForm'
import AddPatientActionBar from './AddPatientActionBar/AddPatientActionBar'
import {useSelector} from 'react-redux'
import {usePostPatientMutation} from '../../features/patients/PatientsApiSlice'
import React from 'react'
import {RootState} from '../../redux/store'

const AddPatientPage = () => {

    const navigate = useNavigate()
    const [postPatient] = usePostPatientMutation()
    const {currentUser} = useSelector((state: RootState) => state.authentication)

    const newPatientInitialState: Patient = {
        firstName: '',
        lastName: '',
        dateOfBirth: null,
        sex: '',
        diagnosis: '',
        notes: '',
        phoneNumber: '',
        status: 'Awaiting Surgery',
        lastAppointment: new Date('2023-01-18T05:00:00.000+00:00'),
        nextAppointment: new Date('2023-01-20T05:00:00.000+00:00'),
        doctor: {
            userId: currentUser.userId
        }
    }

    const [newPatient, setNewPatient] = useState(newPatientInitialState)


    /**
     * Sets the patient's first name.
     * @param fNameInput First name
     */
    const setPatientFirstName = (fNameInput: string) => {
        setNewPatient({...newPatient, firstName: fNameInput})
    }

    /**
     * Sets the patient's last name.
     * @param lNameInput Last name
     */
    const setPatientLastName = (lNameInput: string) => {
        setNewPatient({...newPatient, lastName: lNameInput})
    }

    /**
     * Sets the patient's date of birth
     * @param dateInput a date
     */
    const setPatientDateOfBirth = (dateInput: Date | null) => {
        setNewPatient({...newPatient, dateOfBirth: dateInput})
    }

    /**
     * Sets the patient's sex
     * @param sexInput the patient's sex
     */
    const setPatientSex = (sexInput: string) => {
        setNewPatient({...newPatient, sex: sexInput})
    }

    /**
     * Sets the patient's diagnosis
     * @param diagnosisInput a diagnosis
     */
    const setPatientDiagnosis = (diagnosisInput: string) => {
        setNewPatient({...newPatient, diagnosis: diagnosisInput})
    }

    /**
     * Sets the patient's notes
     * @param notesInput notes
     */
    const setPatientNotes = (notesInput: string) => {
        setNewPatient({...newPatient, notes: notesInput})
    }

    /**
     * Sets the patient's phone number
     * @param phoneNumberInput a phone number
     */
    const setPatientPhoneNumber = (phoneNumberInput: string) => {
        setNewPatient({...newPatient, phoneNumber: phoneNumberInput})
    }

    /**
     * Saves the currentPatient to the patient list and navigates back to the patient register page
     */
    const saveNewPatient = () => {
        postPatient(newPatient)
        navigate('/patients')
    }

    return(
        <div className={styles.page}>
            <div>
             Patient Register {'>'} Add Patient
            </div>
            <AddPatientActionBar onSave={saveNewPatient}/>
            <AddPatientForm
                currentPatient={newPatient}
                onChangeFirstName={setPatientFirstName}
                onChangeLastName={setPatientLastName}
                onChangeDOB={setPatientDateOfBirth}
                onChangeSex={setPatientSex}
                onChangeDiagnosis={setPatientDiagnosis}
                onChangeNotes={setPatientNotes}
                onChangePhoneNumber={setPatientPhoneNumber}
            />
        </div>
    )
}

export default AddPatientPage
