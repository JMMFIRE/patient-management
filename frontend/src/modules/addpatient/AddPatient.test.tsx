import '@testing-library/jest-dom'
import {resetPatientList} from '../../mock/api/handlers/patients-api-handler'
import {loggedInPreloadedState, renderWithProviders} from '../../common/test/test-utils'
import AddPatientPage from './AddPatientPage'
import React from 'react'
import {fireEvent, screen, waitFor} from '@testing-library/react'
import {response} from 'msw'


beforeEach(() => {
    resetPatientList()
})

describe('Add Patients Page', () => {

    it('Renders the add patient page', async () => {
        renderWithProviders(<AddPatientPage/>, loggedInPreloadedState)
        const addPatientForm = await screen.findByTestId('add-patient-form')
        expect(addPatientForm).toBeInTheDocument()
    })

    it('Renders the add patient page, matches snapshot', async () => {
        const addPatientPage = renderWithProviders(<AddPatientPage/>, loggedInPreloadedState)
        expect(addPatientPage).toMatchSnapshot()
    })

    it('Cancels the addition of a new patient', async () => {
        const addPatientPage = renderWithProviders(<AddPatientPage/>, loggedInPreloadedState)
        const cancelButton = await screen.findByTestId('cancel-button')

        fireEvent.click(cancelButton)

        await waitFor(() => expect(addPatientPage.mockedNavigate).toBeCalled())
        expect(addPatientPage.mockedNavigate.mock.lastCall[0]).toBe('/patients')
    })

    it('Adds a new patient', async () => {
        const addPatientPage = renderWithProviders(<AddPatientPage/>, loggedInPreloadedState)

        const firstName = await screen.findByTestId('first-name-field')
        const lastName = await screen.findByTestId('last-name-field')
        const dob = await screen.getByLabelText('Date of birth')
        const maleButton = await screen.findByTestId('male-button')
        const diagnosis = await screen.findByTestId('diagnosis-field')
        const notes = await screen.findByTestId('notes-field')
        const phoneNumber = await screen.findByTestId('phone-number-field')
        const saveButton = await screen.findByTestId('save-button')

        fireEvent.change(firstName, { target: { value: 'Test fName'}})
        fireEvent.change(lastName, { target: { value: 'Test lName'}})
        fireEvent.change(dob, { target: { value: '05/20/1997' }})
        fireEvent.click(maleButton)
        fireEvent.change(diagnosis, { target: { value: 'Test diagnosis'}})
        fireEvent.change(notes, { target: { value: 'Test notes'}})
        fireEvent.change(phoneNumber, { target: { value: 'Test phone number'}})
        fireEvent.click(saveButton)

        await waitFor(() => expect(addPatientPage.mockedNavigate).toBeCalled())
        expect(addPatientPage.mockedNavigate.mock.lastCall[0]).toBe('/patients')

        const { status } = await response.once()

        expect(status).toBe(200)
    })
})
