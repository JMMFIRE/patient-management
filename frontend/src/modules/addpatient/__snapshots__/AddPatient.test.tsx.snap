// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Patients Page Renders the add patient page, matches snapshot 1`] = `
{
  "mockedNavigate": [MockFunction],
  "renderResult": {
    "asFragment": [Function],
    "baseElement": <body>
      <div>
        <div
          class="page"
        >
          <div>
            Patient Register 
            &gt;
             Add Patient
          </div>
          <div
            class="actionBar"
          >
            <div
              class="header"
            >
              Add Patient
            </div>
            <div
              class="actionMenu"
            >
              <div
                class="actions"
              >
                <button
                  class="button btn btn-outline-primary"
                  data-testid="cancel-button"
                  type="button"
                >
                  Cancel
                </button>
                <button
                  class="button btn btn-primary"
                  data-testid="save-button"
                  type="button"
                >
                  Save
                </button>
              </div>
            </div>
          </div>
          <div
            class="formWrapper"
          >
            <form
              class="form"
              data-testid="add-patient-form"
            >
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                >
                  Record number
                </label>
                <div
                  class="col-sm-8"
                >
                  <label
                    class="form-label"
                  >
                    Record number will be assigned automatically when you save.
                  </label>
                </div>
              </div>
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                >
                  First name
                </label>
                <div
                  class="col-sm-6"
                >
                  <input
                    class="form-control"
                    data-testid="first-name-field"
                  />
                </div>
              </div>
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                >
                  Last name
                </label>
                <div
                  class="col-sm-6"
                >
                  <input
                    class="form-control"
                    data-testid="last-name-field"
                  />
                </div>
              </div>
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                  for="date-of-birth"
                >
                  Date of birth
                </label>
                <div
                  class="col-sm-6"
                >
                  <div
                    class="react-datepicker-wrapper"
                  >
                    <div
                      class="react-datepicker__input-container"
                    >
                      <span
                        aria-live="polite"
                        class="react-datepicker__aria-live"
                        role="alert"
                      />
                      <input
                        class="form-control"
                        id="date-of-birth"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                >
                  Sex
                </label>
                <div
                  class="sexButtons col-sm-6"
                >
                  <button
                    class="btn btn-secondary"
                    data-testid="male-button"
                    type="button"
                  >
                    Male
                  </button>
                  <button
                    class="btn btn-secondary"
                    data-testid="female-button"
                    type="button"
                  >
                    Female
                  </button>
                </div>
              </div>
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                >
                  Diagnosis
                </label>
                <div
                  class="col-sm-6"
                >
                  <input
                    class="form-control"
                    data-testid="diagnosis-field"
                  />
                </div>
              </div>
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                >
                  Notes
                </label>
                <div
                  class="col-sm-6"
                >
                  <textarea
                    class="form-control"
                    data-testid="notes-field"
                    style="height: 7rem;"
                  />
                </div>
              </div>
              <div
                class="group row"
              >
                <label
                  class="form-label col-form-label col-sm-4"
                >
                  Phone number
                </label>
                <div
                  class="col-sm-6"
                >
                  <input
                    class="form-control"
                    data-testid="phone-number-field"
                  />
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </body>,
    "container": <div>
      <div
        class="page"
      >
        <div>
          Patient Register 
          &gt;
           Add Patient
        </div>
        <div
          class="actionBar"
        >
          <div
            class="header"
          >
            Add Patient
          </div>
          <div
            class="actionMenu"
          >
            <div
              class="actions"
            >
              <button
                class="button btn btn-outline-primary"
                data-testid="cancel-button"
                type="button"
              >
                Cancel
              </button>
              <button
                class="button btn btn-primary"
                data-testid="save-button"
                type="button"
              >
                Save
              </button>
            </div>
          </div>
        </div>
        <div
          class="formWrapper"
        >
          <form
            class="form"
            data-testid="add-patient-form"
          >
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
              >
                Record number
              </label>
              <div
                class="col-sm-8"
              >
                <label
                  class="form-label"
                >
                  Record number will be assigned automatically when you save.
                </label>
              </div>
            </div>
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
              >
                First name
              </label>
              <div
                class="col-sm-6"
              >
                <input
                  class="form-control"
                  data-testid="first-name-field"
                />
              </div>
            </div>
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
              >
                Last name
              </label>
              <div
                class="col-sm-6"
              >
                <input
                  class="form-control"
                  data-testid="last-name-field"
                />
              </div>
            </div>
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
                for="date-of-birth"
              >
                Date of birth
              </label>
              <div
                class="col-sm-6"
              >
                <div
                  class="react-datepicker-wrapper"
                >
                  <div
                    class="react-datepicker__input-container"
                  >
                    <span
                      aria-live="polite"
                      class="react-datepicker__aria-live"
                      role="alert"
                    />
                    <input
                      class="form-control"
                      id="date-of-birth"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
              >
                Sex
              </label>
              <div
                class="sexButtons col-sm-6"
              >
                <button
                  class="btn btn-secondary"
                  data-testid="male-button"
                  type="button"
                >
                  Male
                </button>
                <button
                  class="btn btn-secondary"
                  data-testid="female-button"
                  type="button"
                >
                  Female
                </button>
              </div>
            </div>
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
              >
                Diagnosis
              </label>
              <div
                class="col-sm-6"
              >
                <input
                  class="form-control"
                  data-testid="diagnosis-field"
                />
              </div>
            </div>
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
              >
                Notes
              </label>
              <div
                class="col-sm-6"
              >
                <textarea
                  class="form-control"
                  data-testid="notes-field"
                  style="height: 7rem;"
                />
              </div>
            </div>
            <div
              class="group row"
            >
              <label
                class="form-label col-form-label col-sm-4"
              >
                Phone number
              </label>
              <div
                class="col-sm-6"
              >
                <input
                  class="form-control"
                  data-testid="phone-number-field"
                />
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>,
    "debug": [Function],
    "findAllByAltText": [Function],
    "findAllByDisplayValue": [Function],
    "findAllByLabelText": [Function],
    "findAllByPlaceholderText": [Function],
    "findAllByRole": [Function],
    "findAllByTestId": [Function],
    "findAllByText": [Function],
    "findAllByTitle": [Function],
    "findByAltText": [Function],
    "findByDisplayValue": [Function],
    "findByLabelText": [Function],
    "findByPlaceholderText": [Function],
    "findByRole": [Function],
    "findByTestId": [Function],
    "findByText": [Function],
    "findByTitle": [Function],
    "getAllByAltText": [Function],
    "getAllByDisplayValue": [Function],
    "getAllByLabelText": [Function],
    "getAllByPlaceholderText": [Function],
    "getAllByRole": [Function],
    "getAllByTestId": [Function],
    "getAllByText": [Function],
    "getAllByTitle": [Function],
    "getByAltText": [Function],
    "getByDisplayValue": [Function],
    "getByLabelText": [Function],
    "getByPlaceholderText": [Function],
    "getByRole": [Function],
    "getByTestId": [Function],
    "getByText": [Function],
    "getByTitle": [Function],
    "queryAllByAltText": [Function],
    "queryAllByDisplayValue": [Function],
    "queryAllByLabelText": [Function],
    "queryAllByPlaceholderText": [Function],
    "queryAllByRole": [Function],
    "queryAllByTestId": [Function],
    "queryAllByText": [Function],
    "queryAllByTitle": [Function],
    "queryByAltText": [Function],
    "queryByDisplayValue": [Function],
    "queryByLabelText": [Function],
    "queryByPlaceholderText": [Function],
    "queryByRole": [Function],
    "queryByTestId": [Function],
    "queryByText": [Function],
    "queryByTitle": [Function],
    "rerender": [Function],
    "unmount": [Function],
  },
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
}
`;
